server/mqtt_ex_test_test.go

```diff
@@ -0,0 +1,569 @@
+
+type mqttExTarget struct {
+	servers  []*Server
+	clusters []*cluster
+	configs  []mqttExTestConfig
+	all      []mqttExDial
+}
+
+type mqttExTestConfig struct {
+	name string
+	pub  []mqttExDial
+	sub  []mqttExDial
+}
+
+func TestMQTTExCompliance(t *testing.T) {
+	if mqttexCLICommandPath == "" {
+		t.Skip(`"mqtt" command is not found in $PATH nor $MQTT_CLI. See https://hivemq.github.io/mqtt-cli/docs/installation/#debian-package for installation instructions`)
+	}
+
+	o := testMQTTDefaultOptions()
+	s := testMQTTRunServer(t, o)
+	o = s.getOpts()
+	defer testMQTTShutdownServer(s)
+
+	cmd := exec.Command(mqttexCLICommandPath, "test", "-V", "3", "-p", strconv.Itoa(o.MQTT.Port))
+
+	output, err := cmd.CombinedOutput()
+	t.Log(string(output))
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			t.Fatalf("mqtt cli exited with error: %v", exitError)
+		}
+	}
+}
+
+func TestMQTTExRetainedMessages(t *testing.T) {
+	if mqttexTestCommandPath == "" {
+		t.Skip(`"mqtt-test" command is not found in $PATH.`)
+	}
+
+	for _, topo := range []struct {
+		name  string
+		makef func(testing.TB) *mqttExTarget
+	}{
+		{
+			name:  "single server",
+			makef: mqttExMakeServer,
+		},
+		{
+			name:  "server with leafnode",
+			makef: mqttExMakeServerWithLeafnode("HUBD", "LEAFD", true),
+		},
+		{
+			name:  "server with leafnode no domains",
+			makef: mqttExMakeServerWithLeafnode("", "", true),
+		},
+		{
+			name:  "server with leafnode no system account",
+			makef: mqttExMakeServerWithLeafnode("HUBD", "LEAFD", false),
+		},
+		{
+			name:  "cluster",
+			makef: mqttExMakeCluster(4, ""),
+		},
+		{
+			name:  "cluster with leafnode cluster",
+			makef: mqttExMakeClusterWithLeafnodeCluster("HUBD", "LEAFD", true),
+		},
+		{
+			name:  "cluster with leafnode cluster no system account",
+			makef: mqttExMakeClusterWithLeafnodeCluster("HUBD", "LEAFD", false),
+		},
+	} {
+		t.Run(topo.name, func(t *testing.T) {
+			target := topo.makef(t)
+			t.Cleanup(target.Shutdown)
+
+			// TODO (levb): if we do not pre-initialize the streams and
+			// consumers, they fail to receive the retained messages in the
+			// configurations with different JetStream domains. Not sure if it
+			// is a bug.
+
+			// initialize the MQTT assets by "touching" all nodes in the
+			// cluster, but then reload to start with fresh server state.
+			for _, dial := range target.all {
+				mqttExInitServer(t, dial)
+			}
+			target.Reload(t)
+
+			numRMS := 100
+			strNumRMS := strconv.Itoa(numRMS)
+			for _, tc := range target.configs {
+				t.Run(tc.name, func(t *testing.T) {
+					topic := "subret_" + nuid.Next()
+
+					// publish numRMS retained messages, one at a time,
+					// round-robin across pub nodes.
+					iNode := 0
+					for i := 0; i < numRMS; i++ {
+						pubTopic := fmt.Sprintf("%s/%d", topic, i)
+						dial := tc.pub[iNode%len(tc.pub)]
+						mqttexRunTest(t, "pub", dial,
+							"--retain",
+							"--topic", pubTopic,
+							"--qos", "0",
```

Comment:
Should it not be `"1"` here to persist a server restart (I tried, and still tests fail, but I still think it should be "1")

Reply:
Actually, if I remember, we always store retained messages, even if they were QoS0. So my comment about qos needed to be "1" is likely wrong.

Reply:
Exactly. We persist them even for QOS0, but it may take longer than publishing them (since the publisher doesn't wait for acks). Originally, I coded a PUBRET single test that used the same QOS, and QOS0 was important to measure the receive performance without the acks.

For this test I broke PUB and SUB into separate steps, to test out delivery after a server restart, but I kept the QOS0 publishing + a simple SUB to wait for all publish's to be processed.